python/EBS_SNAPSHOT_PUBLIC_RESTORABLE_CHECK/EBS_SNAPSHOT_PUBLIC_RESTORABLE_CHECK.py

```diff
@@ -0,0 +1,417 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for
+# the specific language governing permissions and limitations under the License.
+'''
+#####################################
+##           Gherkin               ##
+#####################################
+Rule Name:
+  EBS_SNAPSHOT_PUBLIC_RESTORABLE_CHECK
+Description:
+  Check whether the Amazon EBS snapshots are not publicly restorable. The rule is NON_COMPLIANT if the RestorableByUserIds field is set to 'all'.
+Trigger:
+  Periodic
+Reports on:
+  AWS::::Account
+Rule Parameters:
+  None
+Scenarios:
+  Scenario: 1
+    Given: No snapshots with RestorableByUserIds parameter set to 'all'
+     Then: Return NOT_APPLICABLE
+  Scenario: 2
+    Given: One or more snapshots with RestorableByUserIds parameter set to 'all'
+     Then: Return NON_COMPLIANT with Annotation containing SnapshotIDs
+'''
+import json
+import datetime
+import boto3
+import botocore
+
+##############
+# Parameters #
+##############
+
+# Define the default resource to report to Config Rules
+DEFAULT_RESOURCE_TYPE = 'AWS::::Account'
+
+# Set to True to get the lambda to assume the Role attached on the Config Service (useful for cross-account).
+ASSUME_ROLE_MODE = False
+
+#############
+# Main Code #
+#############
+
+# Function to generate evaluations for all non compliant resources
+def generate_eval_list(snapshots, event):
+    evaluations = []
+    for snapshot in snapshots:  # Looping through all available snapshots
+        evaluation = build_evaluation(snapshot['SnapshotId'], "NON_COMPLIANT", event, resource_type=DEFAULT_RESOURCE_TYPE, annotation="EBS Snapshot: {} is public".format(snapshot['SnapshotId']))
+        evaluations.append(evaluation)
+    return evaluations
+
+# Function to obtain all public snapshots
+def get_public_snapshots(ec2_client, owner_id):
+    snapshots = []
+    next_token = None
+    snapshots_result = {}
+    while next_token is None or next_token:  # Checking if there is another token or its the first invocation
+        try:
+            if next_token is None:  # If next_token is None; make first call
+                snapshots_result = ec2_client.describe_snapshots(OwnerIds=[owner_id], RestorableByUserIds=['all'], MaxResults=1000)
+            else:  # Else use the NextToken to obtain next set of snapshots
+                snapshots_result = ec2_client.describe_snapshots(OwnerIds=[owner_id], NextToken=next_token, MaxResults=1000)
+            if snapshots_result['ResponseMetadata']['HTTPStatusCode'] != 200:  # Check if API call was unsuccessful
+                return(False, snapshots)
+        except Exception as boto3_exception:
+            print(boto3_exception)
+            return(False, snapshots)
+        snapshots.extend(snapshots_result['Snapshots'])
+        if 'NextToken' in snapshots_result:  # If NextToken is present in snapshots_result, assign NextToken for next API call

```

Comment:
You don't need to put comment on every line. Un-necessary comments can be removed. You can keep comments which describes things like what a function or loop is doing.

Reply:
Removed excess comments